### ISSUE TITLE
Support for Delphi 6

### ISSUE THREAD

**Issue by andersmelander** [andersmelander/map2pdb#1]
**[Original report](https://bitbucket.org/anders_melander/map2pdb/issues/1) by Mattes D (Bitbucket: [Mattes D](https://bitbucket.org/Mattes D), ).**

----------------------------------------

The Delphi 6 linker produces MAP files that cannot be read by map2pdb out of the box. It reports overlapping segments via an exception, with no details as to what the offending segments are.

Is there any chance of getting this ancient IDE supported by this tool? It would help me immensely.

I’m attaching a sample Hello World application source, executable and map file produced.

**Comment by andersmelander** [comment]
**Original comment by Anders@Melander (Bitbucket: [Anders](https://bitbucket.org/Anders), GitHub: [Anders](https://github.com/Anders)).**

----------------------------------------

Yes, it’s a bit inconvenient that the error message doesn’t state which the overlapping segments are \(I’ll fix that\) but in your case there’s no doubt - because they all overlap:

```
 Start         Length     Name                   Class
 0001:00000000 000013A0H .text                   CODE
 0002:00000000 0000009CH .data                   DATA
 0002:0000009C 000006E5H .bss                    BSS
```

As you can see the `.data` and `.bss` segments both overlap the `.text` segment. On top of that you have two segments with the same index: `0002`. This is also illegal.

If I recalculate the segment offsets and make the indices unique:

```
 Start         Length     Name                   Class
 0001:00000000 000013A0H .text                   CODE
 0002:000013A0 0000009CH .data                   DATA
 0003:0000143C 00001B21H .bss                    BSS
```

Then the segments pass but now the symbol parser bugs out because the format of the symbol names is different from that of newer Delphi versions \(and there’s also a few bugs in my symbol name parser\). I’ll get the bugs fixed and modify the symbol name parser so it can handle the Delphi 6 format, but I think you will have to handle the invalid segment indices and offset values yourself.

**Comment by andersmelander** [comment]
**Original comment by Anders Melander (Bitbucket: [anders_melander](https://bitbucket.org/anders_melander), ).**

----------------------------------------

No there’s no formal definition of the map file format. It’s not a file format that is consumed by Delphi itself so it’s perfectly possible for Delphi to produce garbage map files and never notice it.

While it might be possible to create a set of rules that will fit your map file, it will mean that large parts of the current design must be rewritten. I rely on segment indices being unique \(it’s the key into the segment collection\) and I rely on segment offsets and size being correct.

If you look at the map files of newer Delphi versions then you will see that no segments overlap and all indices are unique. The same applies for pdb files produced by for example VS.

Here’s the EXE sections of your application:

![](https://bitbucket.org/repo/g5Mq5kb/images/3298187427-billede.png)
As you can see the `.text`, `.data` and `.bss` are three separate sections and they should have been tree separate segments in the map file.

**Comment by andersmelander** [comment]
**Original comment by Anders Melander (Bitbucket: [anders_melander](https://bitbucket.org/anders_melander), ).**

----------------------------------------

Fixes #1
This change enables the map reader to parse Delphi 6 generated map files.
Map reader: Fix for symbol name that does not start with module name causes error. Fix various related string indexing bugs.

**Comment by andersmelander** [comment]
**Original comment by Anders Melander (Bitbucket: [anders_melander](https://bitbucket.org/anders_melander), ).**

----------------------------------------

Added support for legacy map file (e.g Delphi 7) with invalid segment and module info.

Fixes #1
Fixes #7
